gwcs/wcs/_pipeline.py

```diff
@@ -0,0 +1,564 @@
+                Step(output_frame, None),
+            ]
+
+        if isinstance(forward_transform, Model):
+            if output_frame is None:
+                msg = (
+                    "An output_frame must be specified "
+                    "if forward_transform is a model."
+                )
+                raise CoordinateFrameError(msg)
+
+            forward_transform = [
+                Step(input_frame, forward_transform.copy()),
+                Step(output_frame, None),
+            ]
+
+        if not isinstance(forward_transform, list):
+            msg = (
+                "Expected forward_transform to be a None, model, or a "
+                f"(frame, transform) list, got {type(forward_transform)}"
+            )
+            raise TypeError(msg)
+
+        self._extend(forward_transform)
+
+    @property
+    def pipeline(self) -> list[Step]:
+        """
+        Allow direct access to the raw pipeline steps.
+        """
+
+        # TODO: This can still allow direct modification of the pipeline list
+        #       without any of the checks and handling that have been put in
+        #       place in order to ensure the pipeline is functional.
+        #       -> Maybe we should return a copy?
+        return self._pipeline
+
+    @property
+    def available_frames(self) -> list[str]:
+        """
+        List of all the frame names in this WCS in their order in the pipeline
+        """
+        return [step.frame.name for step in self._pipeline]
+
+    def _wrap_step(
+        self, step: Step | StepTuple, *, replace_index: int | None = None
+    ) -> Step:
+        """
+        Wrap the step in a Step object if it is not already, and
+        check that the frame is not already in the pipeline.
+
+        Parameters
+        ----------
+        step : `~gwcs.wcs.Step` or tuple
+            The step to wrap in a Step object and check.
+        replace_index : int or None
+            The index of the step to replace in the pipeline, this ensures that
+            we can inplace replace a step using the same frame as the one being
+            replaced. This frame will be removed from the frames to check against
+            If None (default), do not remove any frames for checking.
+        """
+        # Copy externally created steps to ensure they are not modified outside
+        # the control of the pipeline
+        value = step.copy() if isinstance(step, Step) else Step(*step)
+
+        frames = self.available_frames
+
+        # If we are replacing a step, remove it from the list of frames as we will
+        # not be duplicating it in that case
+        if replace_index is not None:
+            frames.pop(replace_index)
+
+        if value.frame.name in frames:
+            msg = f"Frame {value.frame.name} is already in the pipeline."
+            raise GwcsFrameExistsError(msg)
+
+        # Add the frame as an attribute of the pipeline
+        super().__setattr__(value.frame.name, value.frame)
```

Comment:
I tried to use a `__getattr__` method instead of this but it broke deep copies in some cases for JWST and Romancal regression tests.